### PR TITLE
[top] Add dedicated jtag id package

### PR DESCRIPTION
**Pull request by tjaychen**
- Partially addresses #11616
- allows partners to override jtag id package directly
  without having to modify the seed or generation flow

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
@arnonsha FYI

File: hw/top_earlgrey/jtag_id_pkg.core
Comment (tjaychen) [inline hw/top_earlgrey/jtag_id_pkg.core:11]: i'm doing it this way so that any other block that needs to reference this id doesn't have to worry about the `fileset_partner`.

we should really do the same for `ast_pkg`

Comment (arnonsha) [reply hw/top_earlgrey/jtag_id_pkg.core:11]: Looks good.
We're ready on our side, so you can merge.
Also, there's a foundry CL that needs to be approved as well to support this on our side.

thanks

File: hw/ip/rv_dm/rv_dm.core
Comment (sriyerg) [inline hw/ip/rv_dm/rv_dm.core:11]: RV_DM does not appear to be using this parameter.

File: hw/ip/lc_ctrl/lc_ctrl.core
Comment (sriyerg) [inline hw/ip/lc_ctrl/lc_ctrl.core:11]: LC_CTRL does not appear to use the parameter either. It's probably better for the IPs to not reference this, and let the value be set by the top level instead.

File: hw/top_earlgrey/rtl/jtag_id_pkg.sv
Comment (sriyerg) [inline hw/top_earlgrey/rtl/jtag_id_pkg.sv:9]: nit: The previous code referenced

File: hw/ip/lc_ctrl/lc_ctrl.core
Comment (tjaychen) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: right i added these since the expectation is that rv_dm and lc_ctrl will make use of them eventually.

Comment (arnonsha) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: @tjaychen 
Just reminding about the foundry CL. Please approve it once this PR is merged or our nightly regression tests will fail...

Thanks

Comment (sriyerg) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: So will the top level no longer override the id code value like it is done here? 
https://github.com/lowRISC/opentitan/blob/2023653b220d6fd4dde5fd9055a6ee5eb0802f2b/hw/top_earlgrey/rtl/autogen/top_earlgrey.sv#L2106

Hmm, interestingly `top_earlgrey` currently overrides the value on RV_DM (that too, with `RvDmIdCodeValue`, not `JTAG_IDCODE`), but not the LC_CTRL, which seems like a bug.

Comment (tjaychen) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: i feel like maybe we should change to directly referencing this value inside rv_dm. 
@msfschaffner

Comment (msfschaffner) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: @sriyerg yeah we need to fix that. this is one of the reasons why #11616 got filed.

@tjaychen I am not sure... isn't this a top-level-specific property that we want to propagate down to the block level when instantiating the IPs (because every top-level / chip instance might have a slightly different JTAG config)? I know we already have dependencies from top-level-specific cores down into some of our IPs, but I think it would be cleaner to avoid that in general...

Comment (sriyerg) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: Aah, sorry I didn't realize it was already pointed out. I tend to also think it is better for the top level to drive this value, so the IPs can actually remain oblivious to it (and as a plus, minimize the instances of fusesoc flag based file swapping).

Comment (tjaychen) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: but wouldn't the partners be required to swap the id package file? 
So for a specific chip instance it would indeed be different. 

and yes i think for another top_* directory would have a different value. 
But this ties back to a long standing issue of how we intend to support different tops.

Comment (tjaychen) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: my most preferred approach is honestly to feed this directly from a propreitary module and avoid having to switch parameter declarations.  That's why I originally suggested this be driven from the chip_level_* via AST.

Comment (msfschaffner) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: > but wouldn't the partners be required to swap the id package file? So for a specific chip instance it would indeed be different.
> 
> and yes i think for another top_* directory would have a different value. But this ties back to a long standing issue of how we intend to support different tops.

sorry I may have described this inaccurately. I wasn't suggesting to remove the fusesoc switch, since as you point out partners will need to be able to swap the id package file (`lowrisc:constants:jtag_id_pkg`). what I was suggesting was to only make the top-level design dependent on that id package, and then feed the idcodes defined in that package into `rv_dm` and `lc_ctrl` via an instantiation parameter instead of having `rv_dm` and `lc_ctrl` depend on...

Comment (tjaychen) [reply hw/ip/lc_ctrl/lc_ctrl.core:11]: ah i see. that makes sense. so you mean in the parameter declaration directly reference this?
That is cleaner, let me do that.

Comment (sriyerg) [reply hw/ip/lc_ctrl/lc_ctrl.core]: Right, when I said "minimize the instances of fusesoc flag based file swapping", I meant limit it only to `top_earlgrey.core`.

**Comment by tjaychen** [comment]
@msfschaffner can you double check if this is what you had in mind?

**Review by msfschaffner — APPROVED** [review]
Yes, this is what I had in mind.
One question I still have though is whether the ID codes should be different for `lc_ctrl` and `rv_dm`?

**Review by sriyerg — APPROVED** [review]

**Comment by arnonsha** [comment]
> Yes, this is what I had in mind. One question I still have though is whether the ID codes should be different for `lc_ctrl` and `rv_dm`?

I agree, diffrentiating lc and rv taps makes sense

**Comment by tjaychen** [comment]
> > Yes, this is what I had in mind. One question I still have though is whether the ID codes should be different for `lc_ctrl` and `rv_dm`?
> 
> I agree, diffrentiating lc and rv taps makes sense

that sounds good, but how should these values differ?

**Comment by tjaychen** [comment]
let me merge this first, then when we have an idea of how we want to differentiate i can update this again.

**Pull request merged**
